packages/mattermost-redux/src/reducers/entities/threads/threads.test.js

```diff
@@ -507,28 +511,48 @@ describe('threads', () => {
-            [{id: 't3', last_reply_at: 40, unread_mentions: 0, unread_replies: 1}, {all: ['t1', 't2', 't3'], unread: ['t2', 't3']}],
-            [{id: 't3', last_reply_at: 5, unread_mentions: 0, unread_replies: 0}, {all: ['t1', 't2'], unread: ['t2']}],
-            [{id: 't3', last_reply_at: 5, unread_mentions: 1, unread_replies: 1}, {all: ['t1', 't2'], unread: ['t2']}],
-
-            [{id: 't2', last_reply_at: 40, unread_mentions: 0, unread_replies: 0}, {all: ['t1', 't2'], unread: ['t2']}],
-            [{id: 't2', last_reply_at: 40, unread_mentions: 1, unread_replies: 1}, {all: ['t1', 't2'], unread: ['t2']}],
-            [{id: 't2', last_reply_at: 40, unread_mentions: 0, unread_replies: 1}, {all: ['t1', 't2'], unread: ['t2']}],
-            [{id: 't2', last_reply_at: 5, unread_mentions: 0, unread_replies: 0}, {all: ['t1', 't2'], unread: ['t2']}],
-            [{id: 't2', last_reply_at: 5, unread_mentions: 1, unread_replies: 1}, {all: ['t1', 't2'], unread: ['t2']}],
-        ])('should handle thread %o', (thread, expected) => {
+            ['a', {id: 't3', last_reply_at: 40, unread_mentions: 0, unread_replies: 0}, {a: {all: ['t1', 't2', 't3'], unread: ['t2']}, b: {all: ['t1'], unread: []}}],
+            ['a', {id: 't3', last_reply_at: 40, unread_mentions: 1, unread_replies: 1}, {a: {all: ['t1', 't2', 't3'], unread: ['t2', 't3']}, b: {all: ['t1'], unread: []}}],
+            ['a', {id: 't3', last_reply_at: 40, unread_mentions: 0, unread_replies: 1}, {a: {all: ['t1', 't2', 't3'], unread: ['t2', 't3']}, b: {all: ['t1'], unread: []}}],
+            ['a', {id: 't3', last_reply_at: 5, unread_mentions: 0, unread_replies: 0}, {a: {all: ['t1', 't2'], unread: ['t2']}, b: {all: ['t1'], unread: []}}],
+            ['a', {id: 't3', last_reply_at: 5, unread_mentions: 1, unread_replies: 1}, {a: {all: ['t1', 't2'], unread: ['t2']}, b: {all: ['t1'], unread: []}}],
+            ['a', {id: 't2', last_reply_at: 40, unread_mentions: 0, unread_replies: 0}, {a: {all: ['t1', 't2'], unread: ['t2']}, b: {all: ['t1'], unread: []}}],
+            ['a', {id: 't2', last_reply_at: 40, unread_mentions: 1, unread_replies: 1}, {a: {all: ['t1', 't2'], unread: ['t2']}, b: {all: ['t1'], unread: []}}],
+            ['a', {id: 't2', last_reply_at: 40, unread_mentions: 0, unread_replies: 1}, {a: {all: ['t1', 't2'], unread: ['t2']}, b: {all: ['t1'], unread: []}}],
+            ['a', {id: 't2', last_reply_at: 5, unread_mentions: 0, unread_replies: 0}, {a: {all: ['t1', 't2'], unread: ['t2']}, b: {all: ['t1'], unread: []}}],
+            ['a', {id: 't2', last_reply_at: 5, unread_mentions: 1, unread_replies: 1}, {a: {all: ['t1', 't2'], unread: ['t2']}, b: {all: ['t1'], unread: []}}],
+
+            // DM/GM threads
+            [undefined, {id: 't3', last_reply_at: 40, unread_mentions: 0, unread_replies: 0}, {a: {all: ['t1', 't2', 't3'], unread: ['t2']}, b: {all: ['t1', 't3'], unread: []}}],
```

Comment:
Why aren't we following a similar approach as channels? Channels are either from a team (private or public) or from no team (dms and gms). That way, when we have to query them it is only `thread.team === team || !thread.team`. I foresee this becoming a nightmare with instances with many teams, and potentially giving problems for visibility of dms threads. A way to deal with this is having a key in the state for undefined (i.e. `threadsInTeam: {a: ['t1', 't2'], b: ['t1'], c: [], undefined: ['t3']}`). `undefined` will be transformed to a string to work as a key, and it will be always different to the team id, so it should be safe to use. Another option is to use a specific constant.

Reply:
Not sure that it's a straight forward approach.
when you receive threads inside global threads you get all threads together DM/GM and all Channels. 
Whereas for posts in channels you only get posts in that specific channel which makes it easier to reason/handle.

Reply:
I think that the logic behind channels is similar. When you call getChannelsInTeam (or something like that) you also get DM/GMs, but they are handled different.

I can see that there is some extra complexity in terms of "separating and merging" the DM/GMs from the other threads, but still I think it pays off the redundancy cost your changes are including.

That being said, I don't understand enough of how CRT is architecture on the inside, so feel free to dismiss this idea.

Reply:
The way @larkox described it is how we do it elsewhere, except instead of undefined, we use the empty string. If you dig through the code around the channel sidebar, it shows channels where either `channel.team_id === currentTeamId || channel.team_id === ''`, and the selectors are the ones in charge of that. Similarly, DMs/GMs also have their IDs stored in `state.entities.channels.channelsInTeam['']`.

We don't have to worry about paging channels though so I don't know if that'd have an effect on this piece of code at least.

Reply:
Another difference is that in channels we don't care about ordering, 
but in threads we do, AKA we don't group threads from channels and DM/GMs separately.

That being said, I'll attempt to make this change since it is a very reasonable one.
 I just wanted to point out my considerations first.

Reply:
According to the above we need to do some book-keeping, which is the same as pushing this change.

Reply:
Ah Right, threads have paging unlike channels. That definitely makes this trickier, so this sounds good to me then